src/Runner.Listener/SelfUpdater.cs

```diff
@@ -220,12 +222,25 @@ private async Task DownloadLatestRunner(CancellationToken token)
 
                                 Trace.Info($"Downloading {_targetPackage.DownloadUrl}");
 
-                                using (FileStream fs = new FileStream(archiveFile, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
+                                using (FileStream fs = new FileStream(archiveFile, FileMode.Create, FileAccess.ReadWrite, FileShare.None, bufferSize: 4096, useAsync: true))
                                 using (Stream result = await httpClient.GetStreamAsync(_targetPackage.DownloadUrl))
                                 {
                                     //81920 is the default used by System.IO.Stream.CopyTo and is under the large object heap threshold (85k). 
                                     await result.CopyToAsync(fs, 81920, downloadCts.Token);
                                     await fs.FlushAsync(downloadCts.Token);
+
+                                    // Validate Hash Matches if it is provided
+                                    if (!String.IsNullOrEmpty(_targetPackage.HashValue))
+                                    {
+                                        fs.Seek(0, SeekOrigin.Begin);
+                                        var hash = GetSHA256(fs);
+                                        if (hash != _targetPackage.HashValue)
+                                        {
+                                            // Hash did not match, we can't recover from this, just throw
+                                            throw new Exception($"Computed runner hash {hash} did not match expected Runner Hash {_targetPackage.HashValue} for {_targetPackage.Filename}");
```

Comment:
What will happen after this throw? I would bet the runner will be stuck in an upgrade loop...

Reply:
It tries to download three times then aborts, I can try to short circuit after one if thats preferable.